pkg/urbit/jets/e/aes_siv.c

```diff
@@ -0,0 +1,357 @@
+/* j/5/aes_ecb.c
+**
+*/
+#include "all.h"
+
+#include <openssl/aes.h>
+
+#include "aes_siv.h"
+
+/* functions
+ */
+static void u3r_bytes_reverse(c3_w    a_w,
+                              c3_w    b_w,
+                              c3_y*   c_y,  /* out */
+                              u3_atom d)    /* in */
+{
+  u3r_bytes(a_w, b_w, c_y, d);
+  c3_w i_w;
+  for (i_w = 0; i_w < ((b_w - a_w) / 2) ; i_w++) {
+    c3_y lo = c_y[i_w];
+    c3_y hi = c_y[b_w - i_w - 1];
+    c_y[i_w] = hi;
+    c_y[b_w - i_w - 1] = lo;
+  }
+
+  return;
+}
+
+static u3_noun _siv_en(c3_y* key_y,
+                       c3_w keysize,
+                       u3_atom ads,
+                       u3_atom txt)
+{
+  AES_SIV_CTX* ctx = AES_SIV_CTX_new();
+  if ( 0 == ctx ) {
+    return u3_none;
+  }
+
+  if ( 0 == AES_SIV_Init(ctx, key_y, keysize) ) {
+    AES_SIV_CTX_free(ctx);
+    return u3_none;
+  }
+
+  while (u3_nul != ads) {
+    c3_w ad_w = u3r_met(3, u3h(ads));
+    c3_y* ad_y = u3a_malloc(ad_w);
+    u3r_bytes_reverse(0, ad_w, ad_y, u3h(ads));
+
+    c3_w ret = AES_SIV_AssociateData(ctx, ad_y, ad_w);
+    u3a_free(ad_y);
+
+    if ( 0 == ret ) {
+      AES_SIV_CTX_free(ctx);
+      return u3_none;
+    }
+
+    ads = u3t(ads);
+  }
+
+  c3_w txt_w = u3r_met(3, txt);
+  c3_y* txt_y = u3a_malloc(txt_w);
+  u3r_bytes_reverse(0, txt_w, txt_y, txt);
+
+  c3_w iv_w = 16;
+  c3_y* iv_y = u3a_malloc(iv_w);
+  c3_w out_w = txt_w;
+  c3_y* out_y = u3a_malloc(out_w);
+  if ( 0 == AES_SIV_EncryptFinal(ctx, iv_y, out_y, txt_y, txt_w) ) {
+    u3a_free(iv_y);
+    u3a_free(out_y);
+    u3a_free(txt_y);
+    AES_SIV_CTX_free(ctx);
+    return u3_none;
+  }
+
+  u3a_free(txt_y);
+  AES_SIV_CTX_free(ctx);
+
+  // Read the first 16 bytes as the "iv"
+  u3_noun iv_n = u3i_bytes(16, iv_y);
+  u3_noun msg_n = u3i_bytes(txt_w, out_y);
+
+  // Reverse byte order for output
+  u3_noun rev_iv_n = u3qc_rev(3, iv_w, iv_n);
+  u3_noun rev_msg_n = u3qc_rev(3, txt_w, msg_n);
```

Comment:
It's worth noting that the `+rev` jet leaves much to be desired. It might be better to have a `u3i_bytes_reverse()` here.

Reply:
Also, these could use the `u3k` convention to simply refcounts.

Reply:
I'm going to ignore this given the preexisting usage of rev in a lot of our other jet code.

Reply:
(The u3kc comment otoh, I acted on.)